### ISSUE TITLE
Remove code from documentation and link to example classes

### ISSUE THREAD

**Issue by lwhite1** [jtablesaw/tablesaw#457]
Code in docs is impossible to maintain.

**Comment by benmccann** [comment]
Maybe we could turn each page into a Jupyter notebook.

Here's an example. You can edit it and run it online by clicking "Execute on Binder" from nbviewer. It does take awhile to launch though
https://nbviewer.jupyter.org/github/benmccann/nbviewer-test/blob/master/Tablesaw.ipynb

One caveat with the example - I'm not sure how to get images and tables to properly render in nbviewer yet. I believe it's supported, but haven't been able to investigate

It would give us the following advantages:
- Viewable online with output. User doesn't need to download and run to see what it looks like (assuming we can get the images fixed)
- Supports markdown. Will help make migration easier and allow for rich descriptions
- Can run online to ensure code works.
  - We would need to update periodically to use latest version of Tablesaw. It's relatively easy to identify breakages when upgrading because the code is runnable unlike our current markdown files
  - It's always clear to the user what version is being demonstrated. It may sometimes be on an old version if we haven't updated it, but it shouldn't ever be broken.

**Comment by benmccann** [comment]
I figured out that you have to press "Widgets" > "Save Notebook Widget State" to have it save with the images and tables. The link above works pretty well now!

**Comment by benmccann** [comment]
You can also embed in sphinx as an alternative to nbviewer: https://nbsphinx.readthedocs.io/en/0.4.2/code-cells.html

**Comment by benmccann** [comment]
Something like [Paradox](https://developer.lightbend.com/docs/paradox/current/) which lets you embed Java in Markdown would be another alternative to ensure the docs compile

**Comment by lwhite1** [comment]
Paradox looks great. If the markdown is referencing code in a file that gets compiled and tested regularly, perhaps we're not doomed after all :)

**Comment by ryancerf** [comment]
I put in a half hearted effort trying to get paradox and sbt to work with maven and gave up. 

I came to the conclusion that I would much rather write my own maven plugin that has similar snippet inclusion functionality rather than fiddle with build tools.

I got a proof of concept working pretty quickly and I think it could get the job done (I might also use it for other projects). It copies the docs directory to target and resolves snippets located in source files. It is almost exactly the same as Paradox's snippet inclusion feature, but that is all it does.

Let me know if you are interested and I can try to figure out how to upload it to maven so we could use it.

**Comment by benmccann** [comment]
Ah, yeah, I hadn't really thought about Paradox being primarily available as an SBT plugin.

The Paradox code looks pretty well structured. There's a separate module for the code Paradox stuff and then an SBT module that just has one file to invoke it from SBT. It looks like it'd be pretty feasible to make a Maven module that does the same thing. Most of the code is just passing options through, so I imagine an MVP could be fairly straightforward since you wouldn't need to support all the options from the beginning

See also https://github.com/lightbend/paradox/issues/236

**Comment by benmccann** [comment]
I think it's worth considering. I haven't used either enough to fully understand the tradeoffs, but I wonder if we'd want to develop and maintain a documentation generator long-term. Paradox also offers things like automatic table of contents generation, advanced linking support (e.g. to make it easier to link to [github](https://developer.lightbend.com/docs/paradox/current/directives/linking.html#github-directive) or [javadoc](https://developer.lightbend.com/docs/paradox/current/directives/linking.html#javadoc-directive)), and callouts.

**Comment by benmccann** [comment]
I took a look at creating a Paradox Maven plugin. The tricky thing seems to be that Paradox is written in Scala so you'd want to create the Maven plugin in Scala. I don't know a lot of Scala. I don't think it'd be hard if we could find a good example, but [the only example I found was too outdated to work still](https://github.com/lightbend/paradox/issues/236#issuecomment-524378012)

**Comment by benmccann** [comment]
It's probably easier to use @ryancerf's plugin since it sounds like he already has that working. I filed some requests with the [scala-mojo-support](https://github.com/jsuereth/scala-mojo-support) project, but unless they get addressed I'm not sure I'd want to spend more time on that path

**Comment by ryancerf** [comment]
I do not mind Scala, but mixing build systems has never worked well for me in the past. Let's give my plugin a try (when it is done).

@lwhite1  Hold on to that desire to fix the docs. I should have something soon!

**Comment by benmccann** [comment]
I agree mixing build systems would be a bad idea. I was thinking more to write a Maven plugin that calls the core Scala code, which they've neatly separated from the SBT plugin. But not sure it's worth it since you've already got another solution basically working

**Comment by ryancerf** [comment]
Here is the [plugin](https://github.com/ryancerf/choss-maven-plugin). I think it will work nicely for our limited needs.

I am having some trouble figuring out how to consume it. I tried using jitpack but was not able to get it to work. I dont really want to deal with uploading it to maven at the moment. Do you have any recommendations?

**Comment by benmccann** [comment]
Bintray is the other one I've seen. Uploading to Maven Central isn't too hard except they make you file a JIRA ticket and wait for someone to manually enable you which can make it take awhile to get started.

**Comment by benmccann** [comment]
GitHub just came out with a feature too, but looks like it might be only a waitlist for now? https://github.com/features/package-registry

**Comment by hmf** [comment]
@benmccann 

Just to inform.

> I took a look at creating a Paradox Maven plugin. The tricky thing seems to be that Paradox is written in Scala so you'd want to create the Maven plugin in Scala. 

You don't have to know Scala. You can call Scala directly from  Java. There are some caveats regarding java static classes and methods and type conversion but it's doable. You also have to make sure to add  the Scala run-time library.

See for example: 

* https://lampwww.epfl.ch/~michelou/scala/using-scala-from-java.html
* http://blog.muhuk.com/2016/05/24/how_to_call_scala_from_java__using_scala_classes.html#.XWFTxvwo_Q8

**Comment by benmccann** [comment]
While you can call Scala from Java, it's quite a pain. E.g. a big issue in this case is that [Scala doesn't generate default constructors](https://github.com/lightbend/paradox/issues/236#issuecomment-524378012)

**Comment by hmf** [comment]
@benmccann 
Apologies for the late reply (am on holiday 8-)). 

Looked at the issue you refer to and the the links you provide. Note that the Stackoverflow question you refer to also points to the same link I gave you. There you find an example of setting the default parameters explicitly. 

I may be missing something but why is it *quite a pain* (I think it may be easier than the this [suggestion](https://github.com/lightbend/paradox/issues/236#issuecomment-524691697)). As I see it you have 2 options:
* Use a builder that instantiates the Scala object explicitly setting the default values as shown [here](https://lampwww.epfl.ch/~michelou/scala/using-scala-from-java.html)(last section). 

* Use a facade that delegates to the Scala obtect. Same as above, only now you havs a set of constructors you can use directly. 

If you want to checkout a branch with what you want/need I can try and help.
